[FIXED JENKINS-40252] Add an Iterable<T> that returns all items unsorted

stephenc opened the pull request:
See [JENKINS-40252](https://issues.jenkins-ci.org/browse/JENKINS-40252)

@jenkinsci/code-reviewers @reviewbybees

stephenc committed:
[FIXED JENKINS-40252] Add an Iterable<T> that returns all items unsorted

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -386,6 +387,35 @@ String name(Item i) {
     /**
+     * Gets all the {@link Item}s recursively in the {@link ItemGroup} tree visible to
+     * {@link Jenkins#getAuthentication()} without concern for the order in which items are returned.
+     *
```
Makes sense to mention the iterator is read-inly

oleg-nenashev commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -386,6 +387,35 @@ String name(Item i) {
+     * @param type the type.
+     * @param <T> the type.
+     * @return An {@link Iterable} for all items.
+     * @since FIXME
```
Makes sense to mention the iterator is read-inly

oleg-nenashev replied:
fixed

oleg-nenashev commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -440,6 +470,121 @@ String name(Item i) {
+             */
+            private final Stack<ItemGroup> stack = new Stack<>();
+            /**
+             * The iterator of the current {@link ItemGroup} we
```
Comment is not complete

oleg-nenashev commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -440,6 +470,121 @@ String name(Item i) {
+                        if (stack.isEmpty()) {
+                            return false;
+                        }
+                        ItemGroup group = stack.pop();
```
I have a concern about the following case:
* ItemGroup `A` contains folders `D`, `C`, and `B`
* `C` and `D` are Item groups containing items
* On the first cycle we go through `A` => `D` => (elements of `D`)
* Once there is no Elements of `D` left, the code pops back to the ItemGroup `A`
* The code takes the **new** iterator, so the `delegate` should get `D` as a next item again

Maybe I'm reading the code wrong since the test seem to cover this case

stephenc replied:
No it will exhaust the current iteration before popping the stack, not strictly breath first

oleg-nenashev commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -440,6 +470,121 @@ String name(Item i) {
+                            // slower path because the caller has switched authentication
+                            // we need to keep the original authentication so that allItems() can be used
+                            // like getAllItems() without the cost of building the entire list up front
+                            try (ACLContext ctx = ACL.as(authentication)) {
```
Why don't you impersonate once and only once on the top level of `hasNext()`? I see not so much reason to potentially reauthenticate it multiple times within the method (depends on the three structure).

Calling the impersonation on the `allItems()` level could also improve the performance a bit

stephenc replied:
Because the usage pattern may be different. With getAllItems() you get the list using current auth, you can then switch auth and do whatever you need...

With this, if you change auth *after* getting the Iterable it will behave the same as getting the List. Without the impersonation, we have an unpredictable iterable. 

The fast case where you just use the same auth throughout will not impersonate.

```
for (Item i: Items.allItems(Jenkins.getInstance(), Item.class)) {
  doSomething(i); // has side-effect of changing auth
}
```

Now the side-effect may be legitimate, eg authorise project related... or it may be a bug...

But if we don't guard against it then it is not safe to replace getAllItems with allItems.

The slow part will be the call to getItems() as that does the permission checks, impersonation around that should be relatively cheap

stephenc replied:
For 99.9% of use cases, people will not be changing authentication while iterating, and hence

```
if (Jenkins.getAuthentication() == authentication) {
    delegate = group.getItems().iterator();
}
```

Will be the path taken... which will just be inlined to a call to `SecurityContextHolder.getContext().getAuthentication()`... given that the `getItems()` will be making this call anyway, the only side-effect in terms of performance is that we will have pre-seeded the CPU cache with the object chain for the `SecurityContextHolder.getContext().getAuthentication()` call anyway.

There are some cases where the authentication used to enumerate the items does not match to the current authentication. 

In Core there is one example: https://github.com/jenkinsci/jenkins/pull/2665/files#diff-46dd855f38f94f25c68cbdae7e6ba64bL243 the example is correct:

* User decides to rename / move a job
* The operation runs as the user's permission
* The item's `renameTo` does escalation to ACL.SYSTEM for the parts that require this escalation, but otherwise is running as the user who did the rename.
* Thus, while iterating the iterable, the items should be the items visible to ACL.SYSTEM but the current authentication is not ACL.SYSTEM...

The second half of the branch is to cover this exact case. 

HTH

stephenc commented:
Given that the tests pass, I may see about replacing getAllItems where the order is not important to give core the perf improvements

stephenc committed:
[JENKINS-40252] Address code review comments

stephenc committed:
[JENKINS-40252] Expose utility comparator singletons

stephenc committed:
[JENKINS-40252] Switch to allItems where traversal order is not important

- Also switch in cases where we have a subset that is likely significantly smaller and hence quicker to sort

stephenc commented:
@oleg-nenashev round 2? the reliable CI agrees that the tests pass... just Flaky McFlaky with its

![screen shot 2016-12-15 at 14 21 38](https://cloud.githubusercontent.com/assets/209336/21227419/dafdfde8-c2d1-11e6-8e49-625eb289fcc7.png)

oleg-nenashev commented on core/src/main/java/hudson/util/RunList.java:
```diff
@@ -76,7 +76,14 @@ public RunList(Collection<? extends Job> jobs) {
     }
 
-    private Iterable<R> combine(Iterable<Iterable<R>> runLists) {
+    public static <J extends Job<J,R>, R extends Run<J,R>> RunList<R> fromJobs(Iterable<? extends J> jobs) {
```
Javadoc is missing

oleg-nenashev commented:
Looks good (excepting one minor comment)

stephenc committed:
Oleg wanted Javadoc commend

oleg-nenashev commented:
re-:bee:

daniel-beck commented:
@stephenc This would also be the preferred approach for plugins that perform long-running operations on each item, as it would drastically lessen the chance of operating on since deleted items, right? Disk Usage Plugin comes to mind.

stephenc commented:
@daniel-beck it would be the preferred approach everywhere unless you want the full list of items in *sorted* order.

If you only want a partial list, sorting a smaller list will in general be faster than building the full list in sorted order

If you do not care about the order, not sorting will be faster than sorting

If you only want to find the first, lazy iteration will be faster than full collection

daniel-beck commented on core/src/main/java/hudson/model/Items.java:
```diff
@@ -79,6 +80,44 @@
         }
     };
+    /**
+     * A comparator of {@link Item} instances that uses a case-insensitive comparison of {@link Item#getName()}.
```
Wouldn't a natural sort be better? IIRC a user recently requested this.

I.e. 

    item1
    item2
    item10

stephenc commented:
I have not seen any objections to this... merging
